### PR TITLE
refactor[test]: add some sanity checks to `abi_decode` tests

### PR DESCRIPTION
**Pull request by cyberthirst**
### What I did
- added sanity checks to abi decode tests to ensure that we're never failing on calldatasize
- also split the creation of payload into two parts when calling target contract with low-level msg_call to increase readability
- depends on https://github.com/vyperlang/vyper/pull/4091 because the tests from there were also refactored

### Commit message

```
QOL improvements for abi_decode tests:

- added sanity checks to abi decode tests to ensure that we're never
failing on calldatasize
- also split the creation of payload into two parts when calling target
contract with low-level msg_call to increase readability
```

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: tests/functional/builtins/codegen/test_abi_decode.py [hunk 1]
```diff
@@ -486,11 +490,12 @@ def test_abi_decode_arithmetic_overflow(env, tx_failed, get_contract):
     # test based on GHSA-9p8r-4xp4-gw5w:
     # https://github.com/vyperlang/vyper/security/advisories/GHSA-9p8r-4xp4-gw5w#advisory-comment-91841
     # buf + head causes arithmetic overflow
-    code = """
+    buffer_sz = 32 * 3
```
Comment (charles-cooper) [inline tests/functional/builtins/codegen/test_abi_decode.py:493]: maybe either bufsz or buffer_size?

File: tests/functional/builtins/codegen/test_abi_decode.py [hunk 2]
```diff
@@ -8,6 +8,8 @@
 
 TEST_ADDR = "0x" + b"".join(chr(i).encode("utf-8") for i in range(20)).hex()
 
+MSG_CALL_ABI_OVERHEAD_SZ = 4 + 2 * 32
```
Comment (charles-cooper) [inline tests/functional/builtins/codegen/test_abi_decode.py:11]: shorter - `BUFFER_OVERHEAD`

File: tests/functional/builtins/codegen/test_abi_decode.py [hunk 3]
```diff
@@ -500,15 +505,18 @@ def f(x: Bytes[32 * 3]):
...
+    msg_call_overhead = (method_id("f(bytes)"), 0x20, 0x60)  # tuple head  # parent array length
+
+    data = _abi_payload_from_tuple(msg_call_overhead, MSG_CALL_ABI_OVERHEAD_SZ)
+
+    buffer_payload = (
         # parent payload - this word will be considered as the head of the abi-encoded inner array
         # and it will be added to base ptr leading to an arithmetic overflow
-        2**256 - 0x60,
+        2**256
+        - 0x60,
```
Comment (charles-cooper) [inline tests/functional/builtins/codegen/test_abi_decode.py:516]: weird formatting

Comment (cyberthirst) [reply tests/functional/builtins/codegen/test_abi_decode.py:516]: result of linting

File: tests/functional/builtins/codegen/test_abi_decode.py (hunk 2 above)
Comment (cyberthirst) [reply tests/functional/builtins/codegen/test_abi_decode.py]: resolved in https://github.com/vyperlang/vyper/pull/4096/commits/3925c8b3c0bffb19bf766e8114d19a8c5cef5d7f

File: tests/functional/builtins/codegen/test_abi_decode.py (hunk 1 above)
Comment (cyberthirst) [reply tests/functional/builtins/codegen/test_abi_decode.py]: resolved in: https://github.com/vyperlang/vyper/pull/4096/commits/c64eea647158911a9b9c6c5429a8690dfd55df1c

File: tests/functional/builtins/codegen/test_abi_decode.py (hunk 3 above)
Comment (charles-cooper) [reply tests/functional/builtins/codegen/test_abi_decode.py:516]: can extract the comment to the previous line, then this line will just be `buffer_payload = (2**256 - 0x60,)`

File: tests/functional/builtins/codegen/test_abi_decode.py [hunk 4]
```diff
@@ -543,27 +554,30 @@ def f(x: Bytes[32 * 5]):
         *_replicate(0x03, 2),
     )
 
-    data += _abi_payload_from_tuple(payload)
+    data += _abi_payload_from_tuple(buffer_payload, buffer_size)
```
Comment (charles-cooper) [inline tests/functional/builtins/codegen/test_abi_decode.py:557]: the payload now has another word that was not there previously

Comment (charles-cooper) [reply tests/functional/builtins/codegen/test_abi_decode.py:557]: ah never mind, my bad -- forgot about the signature change

**Review by charles-cooper — APPROVED** [review]
this lgtm. @cyberthirst i made some small formatting changes, please take a look and see if they look good to you

File: tests/functional/builtins/codegen/test_abi_decode.py (hunk 3 above)
Comment (charles-cooper) [reply tests/functional/builtins/codegen/test_abi_decode.py]: https://github.com/vyperlang/vyper/pull/4096/commits/576efa3c23a2d6be40f7875495c3aeb39875f958

**Pull request merged**